airbyte-integrations/bases/base-normalization/normalization/transform_catalog/transform.py

```diff
@@ -139,6 +101,269 @@ def extract_schema(profiles_yml: dict) -> str:
+    tables_in_schema: Dict[str, Set[str]],
+    properties,
+    raw_schema,
+    schema,
+    name,
+    table,
+    parent_hash_id: str,
+    inject_sql: str,
+):
+    # Check properties
+    if not properties:
+        print("Ignoring '{}' nested field from {} because properties list is empty".format(name, "/".join(path)))
+        return
+
+    # Generate JSON Parsing model
+    sql_file_name = normalize_identifier_name("ab1_{}".format(name), integration_type)
+    sql = "select\n  "
+    if parent_hash_id:
+        sql += "{},\n  ".format(parent_hash_id)
+    for field in properties.keys():
+        sql_field = json_extract_property(json_col=json_col, name=field, definition=properties[field], integration_type=integration_type)
+        if sql_field and not is_airbyte_column(field):
+            sql += sql_field + ",\n  "
+    sql += "_airbyte_emitted_at"
+    sql += "\nfrom {}\n".format(table)
+    if inject_sql:
+        sql += inject_sql
+    if len(path) > 1:
+        sql += "-- {} from {}\n  ".format(name, "/".join(path))
+    output_sql_view(output, raw_schema, sql_file_name, sql, path)
+
+    # Generate column typing model
+    previous_sql_file_name = sql_file_name
+    table = "{{ ref('" + previous_sql_file_name + "') }}"
+    sql_file_name = normalize_identifier_name("ab2_{}".format(name), integration_type)
+    sql = "select\n  "
+    if parent_hash_id:
+        sql += "{},\n  ".format(parent_hash_id)
+    for field in properties.keys():
+        sql_field = cast_property_type(name=field, definition=properties[field], integration_type=integration_type)
+        if sql_field and not is_airbyte_column(field):
+            sql += sql_field + ",\n  "
+    sql += "_airbyte_emitted_at"
+    sql += "\nfrom {}".format(table)
+    if len(path) > 1:
+        sql += "-- {} from {}\n  ".format(name, "/".join(path))
+    output_sql_view(output, raw_schema, sql_file_name, sql, path)
+
+    hash_id = quote(f"_airbyte_{name}_hashid", integration_type)
+    # Generate hash_id column model
+    previous_sql_file_name = sql_file_name
+    table = "{{ ref('" + previous_sql_file_name + "') }}"
+    sql_file_name = normalize_identifier_name("ab3_{}".format(name), integration_type)
+    sql = "select\n  *,\n  "
+    if properties:
+        sql += "{{ dbt_utils.surrogate_key([\n    "
+        if parent_hash_id:
+            sql += "'{}',\n  ".format(parent_hash_id)
+        for field in properties.keys():
+            sql_field = safe_cast_to_varchar(name=field, definition=properties[field], integration_type=integration_type)
+            if sql_field and not is_airbyte_column(field):
+                sql += sql_field + ",\n    "
+        sql += "\n  ]) }} as " + hash_id
+    else:
+        sql += "null as " + hash_id
+    sql += "\nfrom {}".format(table)
+    if len(path) > 1:
+        sql += "-- {} from {}\n  ".format(name, "/".join(path))
+    output_sql_view(output, raw_schema, sql_file_name, sql, path)
+
+    # Generate dedup final model
```

Comment:
is this a new feature? can we save it for a separate PR? i think it would be good to keep this as lean as possible since it is already a big change and we have a customer demo tomorrow.

Reply:
It's a small cleanup step a little smarter than `distinct`, it's not deduplicating per primary keys... but sure I can remove it.

the "final" model is still necessary though (it's materializing data)